constant_time_derive/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2023 The MobileCoin Foundation
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, Data, DataEnum, DeriveInput, Fields, GenericParam, Generics};
+
+#[proc_macro_derive(ConstantTimeEq)]
+pub fn constant_time_eq(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    derive_ct_eq(&input)
+}
+// TODO: Check or remove padding and align decorators on the struct
+fn parse_fields(fields: &Fields) -> Result<proc_macro2::TokenStream, &'static str> {
+    match &fields {
+        Fields::Named(fields_named) => {
+            let mut token_stream = quote!();
+            let mut iter = fields_named.named.iter().peekable();
+
+            while let Some(field) = iter.next() {
+                let ident = &field.ident;
+                match iter.peek() {
+                    None => token_stream.extend(quote! { {self.#ident}.ct_eq(&{other.#ident}) }),
+                    Some(_) => {
+                        token_stream.extend(quote! { {self.#ident}.ct_eq(&{other.#ident}) & })
+                    }
+                }
+            }
+            Ok(token_stream)
+        }
+        Fields::Unnamed(_) => {
+            let token_stream = quote! { {self.0}.ct_eq({&other.0}) };
+
+            Ok(token_stream)
+        }
```

Comment:
:wrench: TBH I'm not sure exactly what's going on here, but i think it only looks at the first field
```rust
#[derive(ConstantTimeEq)]
struct UnnamedFields(u8, u8);

#[test]
fn unnamed_fields() {
    let a = UnnamedFields(3, 4);
    let b = UnnamedFields(3, 4);
    let c = UnnamedFields(3, 3);
    assert!(bool::from(a.ct_eq(&b)));
    assert!(bool::from(!a.ct_eq(&c)));
}
```
That test fails with
```console
assertion failed: bool::from(!a.ct_eq(&c))
thread 'unnamed_fields' panicked at 'assertion failed: bool::from(!a.ct_eq(&c))', constant_time_derive/tests/newtype_struct.rs:29:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:64:14
   2: core::panicking::panic
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:111:5
   3: newtype_struct::unnamed_fields
             at ./tests/newtype_struct.rs:29:5
   4: newtype_struct::unnamed_fields::{{closure}}
             at ./tests/newtype_struct.rs:24:21
   5: core::ops::function::FnOnce::call_once
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/ops/function.rs:507:5
   6: core::ops::function::FnOnce::call_once
             at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/ops/function.rs:507:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Reply:
I misunderstood that unnamed structs will be for the FFI types. Code should handle this now.